src/transformers/modeling_blenderbot.py

```diff
@@ -0,0 +1,582 @@
+            dropout=config.dropout,
+            relu_dropout=config.relu_dropout,
+            attention_dropout=config.attention_dropout,
+            padding_idx=self.pad_idx,
+            learn_positional_embeddings=config.learn_positional_embeddings,
+            n_positions=self.n_positions,
+            activation=config.activation,
+        )
+        self.decoder = BlenderbotDecoder(
+            n_heads=config.n_heads,
+            embedding_size=config.embedding_size,
+            ffn_size=config.ffn_size,
+            n_layers=config.n_layers,
+            vocabulary_size=config.vocab_size,
+            embedding=self.embeddings,
+            attention_dropout=config.attention_dropout,
+            dropout=config.dropout,
+            relu_dropout=config.relu_dropout,
+            learn_positional_embeddings=config.learn_positional_embeddings,
+            padding_idx=self.pad_idx,
+            n_positions=self.n_positions,
+            activation=config.activation
+        )
+        self.init_weights()
+
+    def forward(self, input_ids, decoder_input_ids=None, position_ids=None, token_type_ids=None):
+        if input_ids is None:
+            raise ValueError("Your input tensor is None, you should specify it")
+        encoder_out = self.encoder(input_ids, token_type_ids)
+        assert isinstance(encoder_out, tuple)
+
+        def prepare_decoder_inputs(input_ids):
+            prev_output_tokens = input_ids.clone()
+            index_of_eos = (input_ids.ne(self.end_idx).sum(dim=1) - 1).unsqueeze(-1)
+            prev_output_tokens[:, 0] = input_ids.gather(1, index_of_eos).squeeze()
+            prev_output_tokens[:, 1:] = input_ids[:, :-1]
+            return prev_output_tokens
+        if decoder_input_ids is None:
+            decoder_input_ids = prepare_decoder_inputs(input_ids)
+        decoder_output = self.decoder(decoder_input_ids, encoder_out)
+        assert isinstance(decoder_output, tuple)
+        tensor_out, new_inc_state = decoder_output
+        output = F.linear(tensor_out, self.embeddings.weight)
+        # force the start token  probability of generation to be 0.
+        output[:, :, self.start_idx] = neginf(output.dtype)
+        return output
+
+
+def reshape_head(tensor, n_heads, dim_per_head):
+    batch_size, seq_len, _ = tensor.size()
+    tensor = tensor.view(batch_size, seq_len, n_heads, dim_per_head)
+    tensor = tensor.transpose(1, 2).contiguous().view(batch_size * n_heads, seq_len, dim_per_head)
+    return tensor
+
+
+def neginf(dtype: torch.dtype):
+    """
+    Return a representable finite number near -inf for a dtype.(credit to parlai)
+    """
+    if dtype is torch.float16:
+        return -1e20
+    else:
+        return -65504
+
+
+def gelu(tensor):
+    """
+    Compute gelu activation function as defined in Gaussian Error Linear Units (GELUs)
+    """
+    return 0.5 * tensor * (1.0 + torch.erf(tensor / math.sqrt(2.0)))
+
+
+def positional_coding(n_pos, dim, out):
```

Comment:
This looks like sinusoidal positional encoding to me => maybe name it like this?